src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmCodeGenerationServiceImpl.java

```diff
@@ -0,0 +1,202 @@
+					return this.applyGroupNamesForManualNaming(new HashSet<>(germplasmCodeNameBatchRequestDto.getGids()),
+						germplasmCodeNameBatchRequestDto.getGermplasmCodeNameSetting(), germplasmNameTypeDTO.get());
+				} else {
+					// For automatic code naming
+					return this.applyGroupNamesForAutomaticNaming(new HashSet<>(germplasmCodeNameBatchRequestDto.getGids()),
+						germplasmNameTypeDTO.get());
+				}
+			} catch (final Exception e) {
+				throw new ApiRuntimeException("An error has occurred when trying generate code names", e);
+			}
+		}
+		return Collections.emptyList();
+	}
+
+	protected List<GermplasmCodingResult> applyGroupNamesForAutomaticNaming(final Set<Integer> gidsToProcess,
+		final GermplasmNameTypeDTO nameType) throws RuleException {
+
+		final NamingConfiguration namingConfiguration = this.namingConfigurationService.getNamingConfigurationByName(nameType.getName());
+
+		final List<String> executionOrder = Arrays.asList(this.ruleFactory.getRuleSequenceForNamespace(CODING_RULE_SEQUENCE));
+		final CodingRuleExecutionContext codingRuleExecutionContext = new CodingRuleExecutionContext(executionOrder, namingConfiguration);
+		final List<GermplasmCodingResult> assignCodesResultsList = new ArrayList<>();
+
+		for (final Integer gid : gidsToProcess) {
+			final String generatedCodeName = (String) this.rulesService.runRules(codingRuleExecutionContext);
+			assignCodesResultsList.add(this.applyGroupName(gid, nameType, generatedCodeName));
+			codingRuleExecutionContext.reset();
+		}
+
+		return assignCodesResultsList;
+	}
+
+	protected List<GermplasmCodingResult> applyGroupNamesForManualNaming(final Set<Integer> gids, final GermplasmNameSetting setting,
+		final GermplasmNameTypeDTO nameType) throws InvalidGermplasmNameSettingException {
+		final List<GermplasmCodingResult> assignCodesResultsList = new ArrayList<>();
+		final boolean startNumberSpecified = setting.getStartNumber() != null;
+		Integer startNumber = setting.getStartNumber();
+
+		// Call this method to check first if the name settings are valid.
+		this.germplasmNamingService.getNextNameInSequence(setting);
+
+		for (final Integer gid : gids) {
+			// Increment start number of succeeding germplasm processed based on initial start # specified, if any
+			if (startNumberSpecified) {
+				setting.setStartNumber(startNumber++);
+			}
+			final String nameWithSequence = this.germplasmNamingService.generateNextNameAndIncrementSequence(setting);
+			final GermplasmCodingResult result = this.applyGroupName(gid, nameType, nameWithSequence);
+			assignCodesResultsList.add(result);
+		}
+		return assignCodesResultsList;
+	}
+
+	protected GermplasmCodingResult applyGroupName(final Integer gid,
+		final GermplasmNameTypeDTO nameType, final String generatedCodeName) {
+
+		final GermplasmCodingResult result = new GermplasmCodingResult();
+		result.setGid(gid);
+
+		final Germplasm germplasm = this.germplasmService.getGermplasmByGID(gid);
```

Comment:
As mentioned in Middleware PR, as much as possible, let's not expose `Germplasm `object. It seems to me we only need the mgid information and so we can do one-off retrieval of that for all gids to process

Reply:
You can use germplasmService.getGermplasmDtoById

Reply:
Done